File: Python/Array/EvaluateDivision.py
# https://leetcode.com/problems/evaluate-division/

class Solution:
    def calcEquation(self, equations: List[List[str]], values: List[float], queries: List[List[str]]) -> List[float]:
        quot = collections.defaultdict(dict)
        for (num, den), val in zip(equations, values):
            quot[num][num] = quot[den][den] = 1.0
            quot[num][den] = val
            quot[den][num] = 1 / val
        for k in quot:
            for i in quot[k]:
                for j in quot[k]:
                    quot[i][j] = quot[i][k] * quot[k][j]
        return [quot[num].get(den, -1.0) for num, den in queries]

File: C++/HashTable/EvaluateDivision.cpp
// https://leetcode.com/problems/evaluate-division/

class Solution {
public:
    vector<double> calcEquation(vector<vector<string>>& equations, vector<double>& values, vector<vector<string>>& queries) {
        unordered_map<string, unordered_map<string, double>> dict;
        int n = values.size();
        
        for (int i = 0; i < n; i++) {
            string num = equations[i][0], den = equations[i][1];
            dict[num][num] = 1.0;
            dict[den][den] = 1.0;
            dict[num][den] = values[i];
            dict[den][num] = 1 / values[i];
        }
        
        for (auto& i: dict) {
            string k1 = i.first;
            for (auto& j: dict[k1]) {
                for (auto& k: dict[k1]) {
                    dict[j.first][k.first] = dict[j.first][k1] * dict[k1][k.first];
                }
            }
        }
        
        vector<double> ret;
        for (auto& query: queries) {
            if (dict.count(query[0]) == 0 || dict[query[0]].count(query[1]) == 0)  ret.push_back(-1.0);
            else ret.push_back(dict[query[0]][query[1]]);
            
        }
        return ret;
        
    }
};
